ENH: minor speedups in non-tree KernelExplainer

tylerjereddy opened the pull request:
* A few minor simplifications/adjustments that seem to give about 5-10% better performance for non-tree `KernelExplainer` for the scenario described at gh-3943.

* I don't think we have formal `asv`-style benchmarks in this project, but informal testing with 5 trials each (times in seconds) for the reproducer in the above issue (on an ARM Mac):
  - `master`:     52.803, 50.711, 51.503, 50.613, 51.348
  - this branch:  47.557, 47.128, 48.197, 47.651, 47.216

* This is less impactful than then improvements at gh-3944, but also less intrusive, and the improvements should compound with each other.

* I added postdoc @arhall0 as a co-author on the commit, since they did the initial profiling work to find the ~10% bottleneck.

* Since it appears that we only really need to identify the first point of difference between the arrays, it is probably possible to write an early-break algorithm that is more efficient, but that would probably increase complexity/need for compiled backend, so this is just a start to bump things a little bit.

## Checklist

- [ ] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [ ] Unit tests added (if fixing a bug or adding a new feature)

tylerjereddy committed:
ENH: minor speedups in non-tree KernelExplainer

* A few minor simplifications/adjustments that seem to give about 5-10%
better performance for non-tree `KernelExplainer` for the scenario
described at gh-3943.

* I don't think we have formal `asv`-style benchmarks in this project,
but informal testing with 5 trials each (times in seconds)
for the reproducer in the above issue (on an ARM Mac):
- `master`:     52.803, 50.711, 51.503, 50.613, 51.348
- this branch:  47.557, 47.128, 48.197, 47.651, 47.216

* This is less impactful than then improvements at gh-3944, but also
less intrusive, and the improvements should compound with each other.

Co-authored-by: Aaron R Hall <arhall@lanl.gov>

tylerjereddy commented on shap/explainers/_kernel.py:
```diff
@@ -495,7 +495,10 @@ def explain(self, incoming_instance, **kwargs):
         number_types = (int, float, np.number)
         if isinstance(i, number_types) and isinstance(j, number_types):
-            return 0 if np.isclose(i, j, equal_nan=True) else 1
+            return 0 if np.allclose(i, j, equal_nan=True) else 1
+        elif hasattr(i, "dtype") and hasattr(j, "dtype"):
+            if np.issubdtype(i.dtype, np.number) and np.issubdtype(j.dtype, np.number):
+                return 0 if np.allclose(i, j, equal_nan=True) else 1
```
This part of the changes was needed because some "perfectly-normal" NumPy arrays in the testsuite were not caught by the condition above after removing the `frompyfunc` business (which, of course, doesn't actually make CPython code any faster).

Obviously, make sure you/we are comfortable that the current testsuite is robust for this codepath. It did seem to have many failures without this shim, so possibly coverage is "ok."

CloseChoice replied:
Thanks for the comment. I started a new conversation on these lines 'cause I somehow couldn't comment + suggest on the 4 lines on the previous one. Isn't there also a case where i and j are not a subtype of `np.number`? Is there a chance that we return `None` here? I can't think of one but just to be sure could we do:

```suggestion
            return 0 if np.allclose(i, j, equal_nan=True) else 1
        elif hasattr(i, "dtype") and hasattr(j, "dtype"):
            if np.issubdtype(i.dtype, np.number) and np.issubdtype(j.dtype, np.number):
                return 0 if np.allclose(i, j, equal_nan=True) else 1
        return 0 if i == j else 1
```

To have the general comparison as a fallback for all cases where if/else branches might be missing?

CloseChoice commented on shap/explainers/_kernel.py:
```diff
@@ -510,8 +513,7 @@ def varying_groups(self, x):
                     x_group = x_group.todense()
-                num_mismatches = np.sum(np.frompyfunc(self.not_equal, 2, 1)(x_group, self.data.data[:, inds]))
-                varying[i] = num_mismatches > 0
+                varying[i] = self.not_equal(x_group, self.data.data[:, inds])
```
nice improvement

tylerjereddy committed:
MAINT: PR 3983 revisions

* Expand the handling of array/object types in `not_equal()`
based on reviewer feedback. Add a related test that fails
when using the naive `i == j` fallback.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

tylerjereddy commented on shap/explainers/_kernel.py:
```diff
@@ -495,7 +495,13 @@ def explain(self, incoming_instance, **kwargs):
+                return 0 if np.allclose(i, j, equal_nan=True) else 1
+            if np.issubdtype(i.dtype, np.bool_) and np.issubdtype(j.dtype, np.bool_):
+                return 0 if np.allclose(i, j, equal_nan=True) else 1
+            return 0 if all(i == j) else 1
```
I tried to address https://github.com/shap/shap/pull/3983#discussion_r1947839307, but found that I actually needed a slight adjustment to the fallback for array-like inputs with `all()`. I also added custom handling for `bool_`, which seems "ok" with `np.allclose()`.

tylerjereddy commented on tests/explainers/test_kernel.py:
```diff
@@ -360,3 +360,16 @@ def test_kernel_logits_zeros_ones_probs(nsamples):
     np.testing.assert_allclose(sigm(shap_values.values.sum(1) + explainer.expected_value), pred, atol=1e-04)
+
+
+@pytest.mark.parametrize("dt", [np.bool_, np.object_])
```
This test is a bit weird in the sense that it doesn't really "fail before" and "pass after," but it does fail if the fallback I recently added uses `return 0 if i == j else 1` instead of `return 0 if all(i == j) else 1` because for the `np.object_` `dtype` you'll get the typical `ValueError: The truth value of an array with more than one element is ambiguous`.

So it is more of a sniff test for things slipping through the cracks, but it isn't super picky beyond that, and in fact the original returning of `None` noted above in some cases still allows the test to pass. It may be possible to cook the test up in a way that is more stringent than that, but it is at least doing *something*.

tylerjereddy committed:
MAINT, TST: PR 3983 revisions

* Relax the stringency of the numeric closeness check
in `test_explainer_non_number_dtype`, which was failing in CI.

tylerjereddy commented on tests/explainers/test_kernel.py:
```diff
@@ -360,3 +360,16 @@ def test_kernel_logits_zeros_ones_probs(nsamples):
+    rf.fit(X, y)
+    explainer = shap.KernelExplainer(model=rf.predict_proba, data=X, random_state=seed)
+    shap_values = explainer(X)
+    np.testing.assert_allclose(shap_values.values.max(), 0.26548, rtol=1e-2)
```
the `rtol` needed for CI to be happy is rather high; conversely, the assertion here isn't really needed for the purpose of the test, which is mostly to fail if there is fundamental logic issue in `not_equal` like using `i == j` when `all(i == j)` is needed.

CloseChoice reviewed (approved):
LGMT, very nice.
